nucypher/policy/policies.py

```diff
@@ -728,56 +728,55 @@ def generate_policy_parameters(n: int,
 
     def sample_essential(self,
                          handpicked_ursulas: Set[Ursula],
+                         drawing_step: int = 5,
                          learner_timeout: int = 1,
                          timeout: int = 10,
                          discover_on_this_thread: bool = False) -> Set[Ursula]: # TODO #843: Make timeout configurable
 
-        selected_ursulas = set(handpicked_ursulas)
-        quantity_remaining = self.n - len(handpicked_ursulas)
-
-        # Need to sample some stakers
-
         handpicked_addresses = [ursula.checksum_address for ursula in handpicked_ursulas]
         reservoir = self.alice.get_stakers_reservoir(duration=self.duration_periods,
                                                      without=handpicked_addresses)
+
+        quantity_remaining = self.n - len(handpicked_ursulas)
         if len(reservoir) < quantity_remaining:
-            error = f"Cannot create policy with {quantity_remaining} arrangements"
+            error = f"Cannot create policy with {self.n} arrangements"
             raise self.NotEnoughBlockchainUrsulas(error)
 
-        to_check = set(reservoir.draw(quantity_remaining))
+        # Handpicked Ursulas are not necessarily known
+        to_check = list(handpicked_ursulas) + reservoir.draw(quantity_remaining)
+        checked = []
 
         # Sample stakers in a loop and feed them to the learner to check
         # until we have enough in `selected_ursulas`.
 
         start_time = maya.now()
-        new_to_check = to_check
 
         while True:
 
             # Check if the sampled addresses are already known.
             # If we're lucky, we won't have to wait for the learner iteration to finish.
-            known = {x for x in to_check if x in self.alice.known_nodes}
-            to_check = to_check - known
-
-            known = random.sample(known, min(len(known), quantity_remaining)) # we only need so many
-            selected_ursulas.update([self.alice.known_nodes[address] for address in known])
-            quantity_remaining -= len(known)
+            checked = [x for x in to_check if x in self.alice.known_nodes]
+            to_check = [x for x in to_check if x not in self.alice.known_nodes]
 
-            if quantity_remaining == 0:
+            if len(checked) >= self.n:
```

Comment:
Since `checked` is a list, it's potentially possible for the same node to be in the list several times (e.g., if the reservoir draws the same node twice), which can have negative side-effects on the check on this line. I see that previously we had a set and you changed it into a list, is there a particular motivation for this?

Reply:
Reservoir can't draw the same node twice, it's in its contract to draw without replacement. The list is there to ensure first drawn and available nodes have the priority in case we over-drew and suddenly have too many nodes in the `known` list.